rfcs/0001-subgraph-composition.md

```diff
@@ -0,0 +1,233 @@
+
+## Terminology
+
+The feature is referred to by _query-time subgraph composition_, short:
+_subgraph composition_. 
+
+Terms introduced and used in this RFC:
+
+- _Foreign schema_: The schema of another subgraph from which types are
+  imported.
+- _Imported type_: An entity type imported from another subgraph schema. 
+- _Extended type_: An entity type imported from another subgraph schema and
+  extended in the subgraph that imports it.
+
+## Detailed Design
+
+The following sections make the following assumptions:
+
+- There is an `ethereum/mainnet` subgraph with an `Address` entity type.
+
+### Composing subgraphs by importing types
+
+In order to reference entity types from a foreign schema, a developer would
+first import these entity types from the other subgraph.
+
+Let's say a DAO subgraph contains a `Proposal` type that has a `proposer` field
+that should link to an Ethereum address (think: Ethereum accounts or contracts)
+and a `transaction` field that should link to an Ethereum transaction. The
+developer would then write the DAO subgraph schema as follows:
+
+```graphql
+type _Schema_
+  @import(
+    types: ["Address", { name: "Transaction", as: "EthereumTransaction" }],
+    from: { name: "ethereum/mainnet" }
+  )
+
+type Proposal @entity {
+  id: ID!
+  proposer: Address!
+  transaction: EthereumTransaction!
+}
+```
+
+This would then allow queries that follow the references to addresses and
+transactions, like
+
+```graphql
+{
+  proposals { 
+    proposer { balance address }
+    transaction { hash block { number } }
+  }
+}
+```
+
+### Extending types from foreign schemas
+
+Extending types from another subgraph involves a few steps:
+
+1. Importing the foreign entity types.
+2. Extending these entity types with custom fields.
+3. Managing (e.g. creating) extended entities in subgraph mappings.
+
+Types can be imported either from a subgraph name or from a subgraph ID.
+Importing from a subgraph name means that the exact version of the imported
+subgraph will be identified at query time and its schema may change in arbitrary
+ways over time. Importing from a subgraph ID guarantees that the schema will
+never change but also means that the import points to a subgraph version that
+may become outdated over time.
+
+Let's say the DAO subgraph wants to extend the Ethereum `Address` type to
+include the proposals created by each respective account. To achieve this, the
+developer would write the following schema:
+
+```graphql
+type _Schema_
+  @import(
+    types: ["Address"],
+    from: { name: "ethereum/mainnet" }
+  )
+
+type Proposal @entity {
+  id: ID!
+  proposer: Address!
+}
+
+extend type Address {
+  proposals: [Proposal!]! @derivedFrom(field: "proposal")
+}
+```
+
+This makes queries like the following possible, where you can go "back" from
+accounts and contracts to proposal entities, despite the Ethereum `Address` type
+originally being defined in the `ethereum/mainnet` subgraph.
+
+```graphql
+{
+  accounts {
```

Comment:
I don't follow this example, this queries `accounts` but I don't see an `Account` type anywhere.

Reply:
Oh, yeah, the query is outdated. This should `addresses`. I'll update it.